src/system/SystemLayer.cpp

```diff
@@ -282,28 +282,102 @@ void Layer::StartTimer(uint32_t aMilliseconds, chip::Callback::Callback<> * aCal
  *   @param[in]  aComplete     A pointer to the function called when timer expires.
  *   @param[in]  aAppState     A pointer to the application state object used when timer expires.
  *
- *   @return CHIP_SYSTEM_NO_ERROR On success.
- *   @return CHIP_SYSTEM_ERROR_NO_MEMORY If a timer cannot be allocated.
- *   @return Other Value indicating timer failed to start.
+ *   @return Timer object handle On success.
+ *   @return nullptr if a timer failed to start.
  *
  */
-Error Layer::StartTimer(uint32_t aMilliseconds, TimerCompleteFunct aComplete, void * aAppState)
+Timer * Layer::StartTimer(uint32_t aMilliseconds, TimerCompleteFunct aComplete, void * aAppState)
 {
     Error lReturn;
     Timer * lTimer;
 
     this->CancelTimer(aComplete, aAppState);
     lReturn = this->NewTimer(lTimer);
-    SuccessOrExit(lReturn);
+    if (lReturn != CHIP_SYSTEM_NO_ERROR)
+    {
+        return nullptr;
+    }
 
     lReturn = lTimer->Start(aMilliseconds, aComplete, aAppState);
     if (lReturn != CHIP_SYSTEM_NO_ERROR)
     {
         lTimer->Release();
+        return nullptr;
     }
 
-exit:
-    return lReturn;
+    return lTimer;
+}
+
+/**
+ * @brief
+ *   This method extends a one-shot timer.
+ *
+ *   @param[in]  aMilliseconds Extension time in milliseconds.
+ *   @param[in]  aTimerPtr     A handle to the existing Timer Object to extend.
+ *
+ *   @return CHIP_SYSTEM_NO_ERROR On success.
+ *   @return CHIP_SYSTEM_ERROR_TIMER_NOT_FOUND If the existing timer cannot be found.
+ *   @return CHIP_SYSTEM_ERROR_UNEXPECTED_STATE If System Layer is not initialized.
+ *   @return CHIP_SYSTEM_ERROR_UNEXPECTED_EVENT If the specified timer is overdue.
+ *   @return Other Value indicating timer failed to extend.
+ *
+ */
+Error Layer::ExtendTimer(uint32_t aMilliseconds, Timer * aTimerPtr)
+{
+    VerifyOrReturnError(this->State() == kLayerState_Initialized, CHIP_SYSTEM_ERROR_UNEXPECTED_STATE);
+
+    Timer * lTimer = nullptr;
+    for (size_t i = 0; i < Timer::sPool.Size(); ++i)
+    {
+        Timer * timer = Timer::sPool.Get(*this, i);
+
+        if (timer != nullptr && timer == aTimerPtr)
+        {
+            lTimer = timer;
+            break;
+        }
+    }
+
+    VerifyOrReturnError(lTimer != nullptr, CHIP_SYSTEM_ERROR_TIMER_NOT_FOUND);
+
+    Timer::Epoch currentEpoch = Timer::GetCurrentEpoch();
+
+    // Make sure the currently-running timer is not overdue. The platform timer API has MSEC resolution, we
+    // only extend the timer if it has more than 1 msec to expire.
```

Comment:
I don't understand this 1 msec part.  Are we trying to not extend timers that might have actually expired?  Or are we just trying to avoid the integer overflow issue?  We don't need to do the +1 to deal with integer overflow.

Reply:
The new ExtendTimer API only works for existing active timer and errors out otherwise. +1 msec can prevent the current existing timer become expired during the processing of this function. With this check, even this function get called right before the next tick, the passed-in timer handle is guaranteed not to be removed from the list of active timers in the layer object.